// tc-server-auth/index.js
var express = require("express");
var router = express.Router();
var bodyParser = require('body-parser');
var jwt = require('jsonwebtoken');
var expressJwt = require('express-jwt');
var params = require('express-params');
var passport = require('passport');
var Auth0Strategy = require('passport-auth0');
var config = require('config');
var request = require('request');

module.exports = function(app) {

	app.use(passport.initialize());
	app.use(bodyParser.json());
	params.extend(app);

	var auth0Domain =  config.get('auth0.Domain');
	var auth0Client = config.get('auth0.Client');
	var auth0Secret = config.get('auth0.Secret');

	var tcAPI = config.get('tcAPI');

	//throw error if auth domain/client/secret is not set
	// if (!auth0Domain) throw new Error('Auth0 Domain not configured. Set `TC_AUTH0_DOMAIN` as an environment variable.');
	// if (!auth0Client) throw new Error('Auth0 Client not configured. Set `TC_AUTH0_CLIENT` as an environment variable.');
	// if (!auth0Secret) throw new Error('Auth0 Secret not configured. Set `TC_AUTH0_SECRET` as an environment variable.');

	var strategy = new Auth0Strategy({
	    domain:       auth0Domain,
	    clientID:     auth0Client,
	    clientSecret: auth0Secret,
	    callbackURL:  '/_auth_/callback'
	  }, function(accessToken, refreshToken, extraParams, profile, done) {
	    // accessToken is the token to call Auth0 API (not needed in the most cases)
	    // extraParams.id_token has the JSON Web Token
	    // profile has all the information from the user
	    return done(null, profile, extraParams);
	});

	passport.use(strategy);

	var jwtCheck = expressJwt({ secret: new Buffer(auth0Secret, 'base64') });

	//TODO(DG: 11/10/2014): Move this to config
	//Topcoder handle
  var userWhitelist = [
    'dayal',
    'rockabilly',
    'kbowerma',
    ' _indy',
    'appiriowes'
  ];

  function whitelist(req, res, next) {
  	var authDisabled = config.get('authDisabled');
  	console.log('checking which whitelist to use', authDisabled)
  	if (authDisabled && (authDisabled == 'true')) {
  		mockWhitelistCheck(req, res, next);
  	} else {
			whitelistCheck(req, res, next);
  	}
  }

  function mockWhitelistCheck(req, res, next) {
  	console.log('in mockWhitelistCheck')
    var tcUser = {
    	id: parseInt(req.headers['x-uid']),
    	name: req.headers['x-name'],
    	handle: req.headers['x-handle'],
    	picture: req.headers['x-picture']
    };
    req.tcUser = tcUser;

  	next();
  }

	function whitelistCheck(req, res, next) {
		console.log('in whitelistCheck')
		var whitelistEnabled = false;
		//TODO: verify req.user exists 1st
		//req.user.user_id is the 'userid' from auth0
		console.log('req.user', req.user);
		console.log('req.user.sub', req.user.sub);
		console.log('req.user.user_id', req.user.user_id);
		if (!req.user || !req.user.sub) {
			console.log('LC: Unauthorized - no user', req.user);
			res.status(401).send('LC: Unauthorized - no user')
		}
		else {
			//Use /v2/user/identity to get tc userid, name, handle, pic
			//http://docs.tcapi.apiary.io/#userprofile
			//console.log('going to make call to tc api')
			request(tcAPI + '/v2/user/tcid/' + req.user.sub, function (error, response, body) {
				console.log('back from tc api', error, response.statusCode)

				//TODO(DG: 11/20/2014): improve/cleanup error handling
			  if (!error && response.statusCode == 200) {
			    body = JSON.parse(body);
			    console.log('whitelistEnabled', whitelistEnabled)
					if (whitelistEnabled && userWhitelist.indexOf(body.handle) < 0) {
						//console.log('not in whitelist, redirect: ', body.handle)
						res.status(401).send('LC: Unauthorized')
					}
					else {
				    var tcUser = {
				    	id: body.uid,
				    	name: req.user.name,
				    	handle: body.handle,
				    	picture: req.user.picture
				    };
				    req.tcUser = tcUser;
						next();
					}
			  }
			  else {
			  	//TODO: handle error response from tc api
			  	res.status(503).send('TC API Unavailable')
			  }
			})
		}

	}

	//TODO: Move this API to User Service
	router.route('/_api_/user')
		.all(jwtCheck)
		.all(whitelist)
		.get(function(req, res) {
			res.send(req.tcUser);
		});
	router.route('/_api_/config')
		.get(function(req, res) {
			console.log('here...')
	    var tcWWW = config.get('tcWWW');
	    var tcBaseProjectUrl = config.get('tcProjectBase');
	    var data = {
	        "baseUrl": tcWWW,
	        "baseChallengeDetailsUrl": tcWWW + '/challenge-details/',
	        "baseMemberProfileUrl": tcWWW + '/member-profile/',
	        "tcBaseProjectUrl": tcBaseProjectUrl
	      }
	    res.json(data);

		});


  app.use(router);
  //TODO: enable this if we want generic error handling
	// app.use(function (err, req, res, next) {
	//   if (err.name === 'UnauthorizedError') {
	//     res.status(401).send('invalid token...');
	//   }
	//   next(err);
	// });

	app.get('/_auth_/callback', function(req, res, next) {

	  passport.authenticate('auth0', function(err, user, info) {
	    if (err) return next(err);
	    if (!user) {
	    	return res.redirect('/login');
	    }
	    else {
	    	//console.log('user back from auth', user)
		    req.logIn(user, function(err) {
		    	//console.log('user back from auth 2', user)
		    //   if (err) { return next(err); }
		      // return res.redirect('/users/' + user.username);
		    });
		    //console.log('info: ', info.id_token);
		    //AJAX call for user; https://qa.topcoder.com/v2/user/tcid/
	      return res.redirect('/#/_auth_/login?jwt=' + info.id_token + '&state=' + req.query.state);
	    }
	  })(req, res, next);
	});

	app.get('/_auth_/handleAuthError', function(req, res) {
		console.log('There was an error in authorizing the user');
	})

	return jwtCheck;
}
